src/PLCTag.jl

```diff
@@ -1,81 +1,97 @@
-			global plc_tag_write = Cfunction{Cint, Tuple{Int32, Cint}}(library, :plc_tag_write)
-			global plc_tag_get_size = Cfunction{Cint, Tuple{Int32}}(library, :plc_tag_get_size)
-			global plc_tag_get_uint64 = Cfunction{UInt64, Tuple{Int32, Cint}}(library, :plc_tag_get_uint64)
-			global plc_tag_set_uint64 = Cfunction{Cint, Tuple{Int32, Cint, UInt64}}(library, :plc_tag_set_uint64)
-			global plc_tag_get_int64 = Cfunction{Int64, Tuple{Int32, Cint}}(library, :plc_tag_get_int64)
-			global plc_tag_set_int64 = Cfunction{Cint, Tuple{Int32, Cint, Int64}}(library, :plc_tag_set_int64)
-			global plc_tag_get_uint32 = Cfunction{UInt32, Tuple{Int32, Cint}}(library, :plc_tag_get_uint32)
-			global plc_tag_set_uint32 = Cfunction{Cint, Tuple{Int32, Cint, UInt32}}(library, :plc_tag_set_uint32)
-			global plc_tag_get_int32 = Cfunction{Int32, Tuple{Int32, Cint}}(library, :plc_tag_get_int32)
-			global plc_tag_set_int32 = Cfunction{Cint, Tuple{Int32, Cint, Int32}}(library, :plc_tag_set_int32)
-			global plc_tag_get_uint16 = Cfunction{UInt16, Tuple{Int32, Cint}}(library, :plc_tag_get_uint16)
-			global plc_tag_set_uint16 = Cfunction{Cint, Tuple{Int32, Cint, UInt16}}(library, :plc_tag_set_uint16)
-			global plc_tag_get_int16 = Cfunction{Int16, Tuple{Int32, Cint}}(library, :plc_tag_get_int16)
-			global plc_tag_set_int16 = Cfunction{Cint, Tuple{Int32, Cint, Int16}}(library, :plc_tag_set_int16)
-			global plc_tag_get_uint8 = Cfunction{UInt8, Tuple{Int32, Cint}}(library, :plc_tag_get_uint8)
-			global plc_tag_set_uint8 = Cfunction{Cint, Tuple{Int32, Cint, UInt8}}(library, :plc_tag_set_uint8)
-			global plc_tag_get_int8 = Cfunction{Int8, Tuple{Int32, Cint}}(library, :plc_tag_get_int8)
-			global plc_tag_set_int8 = Cfunction{Cint, Tuple{Int32, Cint, Int8}}(library, :plc_tag_set_int8)
-			global plc_tag_get_float64 = Cfunction{Cdouble, Tuple{Int32, Cint}}(library, :plc_tag_get_float64)
-			global plc_tag_set_float64 = Cfunction{Cint, Tuple{Int32, Cint, Cdouble}}(library, :plc_tag_set_float64)
-			global plc_tag_get_float32 = Cfunction{Cfloat, Tuple{Int32, Cint}}(library, :plc_tag_get_float32)
-			global plc_tag_set_float32 = Cfunction{Cint, Tuple{Int32, Cint, Cfloat}}(library, :plc_tag_set_float32)
+			result = new{T}(Ref{Int32}(tag), count)
+			finalizer(t -> C.plc_tag_destroy(t[]), result.tag)
+			return result
+		end
+	end
+	
+	
+	Base.length(ref::PLCRef) = ref.count
+	Base.size(ref::PLCRef) = (ref.count,)
+	Base.eltype(ref::PLCRef{T}) where {T} = T
+	Base.iterate(ref::PLCRef, state = 1) = state > length(ref) ? nothing : (ref[state], state+1)
+	
+	function Base.getindex(ref::PLCRef, ind::Int = 1)
+		1 <= ind <= length(ref) || error("Index $(ind) is out of range for PLCRef of count $(ref.count)")
+		
+		code = C.plc_tag_read(ref.tag[], DATA_TIMEOUT)
+		code == C.PLCTAG_STATUS_OK || error("ERROR: Unable to read the data! Got error code $(code): $(unsafe_string(C.plc_tag_decode_error(code)))")
```

Comment:
I don't think we want to do this. I think we'll want the async functionality of libplctag, which is invoked by setting `DATA_TIMEOUT` to `0` on `plc_tag_read`. Here is his example of how it's done: https://github.com/kyle-github/libplctag/blob/master/src/examples/async.c

I wrote up a test to see the performance difference on the PLC (I'm at Erico rn):
```julia
using PLCTag

const TAG_PATH = "protocol=ab_eip&gateway=192.168.1.1&path=1,0&cpu=compactlogix&elem_size=4&elem_count=1&name=topTerm&debug=1"
const DATA_TIMEOUT = 5000

const FLOAT_TAG_PATH     = "protocol=ab_eip&gateway=192.168.1.1&path=1,0&cpu=compactlogix&elem_size=4&elem_count=1&name=floater&debug=1"
const FLOAT_DATA_TIMEOUT = 5000

tag = PLCTag.C.plc_tag_create(TAG_PATH, DATA_TIMEOUT)
tag <= 0 && error("ERROR $(PLCTag.C.plc_tag_decode_error(tag)): Could not create tag!")
code = PLCTag.C.plc_tag_status(tag)
code == PLCTag.C.PLCTAG_STATUS_OK || error("ERROR: Error setting up tag internal state. Got error code $(code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(code)))")

float_tag = PLCTag.C.plc_tag_create(FLOAT_TAG_PATH, DATA_TIMEOUT)
float_tag <= 0 && error("ERROR $(PLCTag.C.plc_tag_decode_error(float_tag)): Could not create tag!")
code = PLCTag.C.plc_tag_status(float_tag)
code == PLCTag.C.PLCTAG_STATUS_OK || error("ERROR: Error setting up tag internal state. Got error code $(code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(code)))")



function read_sync()
	@time begin
	# Do 10 reads
		for i in 1:100
			code = PLCTag.C.plc_tag_read(tag, 5000)
			code == PLCTag.C.PLCTAG_STATUS_OK || error("ERROR: Unable to read the data! Got error code $(code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(code)))")
			float_code = PLCTag.C.plc_tag_read(float_tag, 5000)
			float_code == PLCTag.C.PLCTAG_STATUS_OK || error("ERROR: Unable to read the data! Got error code $(float_code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(float_code)))")


			val = PLCTag.C.plc_tag_get_int32(tag, 0)
			float_val = PLCTag.C.plc_tag_get_float32(float_tag, 0)
			val == 1 || error("Did not get 1")
			isapprox(float_val, 1.123) || error("Did not get 1.123")
		end
	end
end

function read_async()
	@time begin
		# Do 10 reads
		for i in 1:100
			code = PLCTag.C.plc_tag_read(tag, 0)
			float_code = PLCTag.C.plc_tag_read(float_tag, 0)

			code == PLCTag.C.PLCTAG_STATUS_OK || code == PLCTag.C.PLCTAG_STATUS_PENDING || error("ERROR: Unable to read the data! Got error code $(code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(code)))")
			
			float_code == PLCTag.C.PLCTAG_STATUS_OK || code == PLCTag.C.PLCTAG_STATUS_PENDING || error("ERROR: Unable to read the data! Got error code $(float_code): $(unsafe_string(PLCTag.C.plc_tag_decode_error(float_code)))")
	

			while true
				if PLCTag.C.plc_tag_status(tag) == PLCTag.C.PLCTAG_STATUS_OK && PLCTag.C.plc_tag_status(float_tag) == PLCTag.C.PLCTAG_STATUS_OK
					break
				end
				sleep(.001)
			end

	
			val = PLCTag.C.plc_tag_get_int32(tag, 0)
			float_val = PLCTag.C.plc_tag_get_float32(float_tag, 0)
			val == 1 || error("Did not get 1")
			isapprox(float_val, 1.123) || error("Did not get 1.123")
		end
	end
end
```

This is only using 2 tags.

As you can see, using async is twice as fast:
```
julia> read_async()
  0.493761 seconds (5.45 k allocations: 123.063 KiB)

julia> read_sync()
  0.891796 seconds (4.30 k allocations: 92.188 KiB)

julia> read_async()
  0.451472 seconds (5.38 k allocations: 121.359 KiB)

julia> read_sync()
  0.852679 seconds (4.30 k allocations: 92.188 KiB)

julia> read_async()
  0.479070 seconds (5.44 k allocations: 122.891 KiB)

julia> read_sync()
  0.959426 seconds (4.30 k allocations: 92.188 KiB)

julia> read_async()
  0.442862 seconds (5.33 k allocations: 120.656 KiB)

julia> read_sync()
  0.847073 seconds (4.30 k allocations: 92.188 KiB)

julia> read_async()
  0.484797 seconds (5.36 k allocations: 121.016 KiB)

julia>
```

I'm guessing the performance will be even more dramatic for many tags. For example, 10 tags might be the same timing for async as 2 tags if network is driving performance.

So I think this abstraction is oversimplifying their api and we'll want async functionality.

Reply:
Yeah, that is good information, we can probably switch to using Futures for getting/setting values asynchronously then.  It probably has something to do with their internal threads and locking in libplctag.

Reply:
Yikes, please test how async scales with number of tags.  Its starting at 0.5 seconds to query 100 times, hopefully doing 10 tags is still 0.5 seconds and not 5x more.

Reply:
yeah, we should have tests for all data types and also performance tests for 1000 tags or something. They are setting up a computer with RSLogix so I'll be able to set up a test tag project to load onto the PLC.

Reply:
I also did these tests over a WiFi network, that probably slowed things down